### PR TITLE
Optimize storage

### PR DESCRIPTION
**Pull request by dorin-iancu**
- keep only 2 storage entries per user instead of `current_week` amount of storage entries.
- added automatic migration for user flags
- added only-owner endpoint for clearing old storage flags

### CODE REVIEW

**Review by claudiu725 — DISMISSED** [review]

File: metabonding/src/claim.rs [hunk 1]
```diff
@@ -81,11 +115,24 @@ pub trait ClaimModule:
             });
         }
 
-        let mut weekly_rewards = ManagedVec::new();
-        for (id, project) in self.projects().iter() {
-            let mut opt_rewards_for_project = None;
+        self.claim_progress_grace_weeks(caller)
```
Comment (ccorcoveanu) [inline metabonding/src/claim.rs]: Side effects in a `validate_and_collect` function, I would refactor this in a separate function. Even `validate_and_collect` is too much in my opinion

File: metabonding/src/claim.rs [hunk 2]
```diff
@@ -35,44 +41,72 @@ pub trait ClaimModule:
...
             );
 
-            let checkpoint: RewardsCheckpoint<Self::Api> = self.rewards_checkpoints().get(week);
+            if grace_weeks_progress.is_week_valid(week) {
+                require!(
+                    grace_weeks_progress.can_claim_for_week(week),
+                    ALREADY_CLAIMED_ERR_MSG
+                );
+
+                grace_weeks_progress.set_claimed_for_week(week);
```
Comment (ccorcoveanu) [inline metabonding/src/claim.rs]: Side-effects in a function that does not suggest this would happen (like also mentioned below)

**Review by ccorcoveanu — CHANGES_REQUESTED** [review]
Functionality-wise this PR looks good, but in my opinion we should refactor the code here before merging this PR since it becomes spaghetti-ish and hard to follow.
I would suggest:
1. No side-effects/state-changes in get functions.
2. If possible refactoring functions that do too many things. Where we have `do_this_AND_that` it should be an indication.
3. We should try to get rid of some nesting - I don't even like two levels but could be acceptable in some places - definitely we should simplify where we have 3 or more levels (we have situations with for loops having if/elseif/else than other for loops inside).
4. In my opinion we should kind-of avoid else branches. Not necessary, but let's try to at least refactor where we have if/elseif/else. Functions with:
```
if something {
  return [...]
}

if something_else {
  return [...]
}

return the else
```
^ become easier to follow than:

```
if something {
  [...]
} else if something_else {
  [...]
} else {
  [...]
}

File: metabonding/src/claim_progress.rs [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    #[storage_mapper("rewardsClaimed")]
+    fn legacy_rewards_claimed_flag(
+        &self,
+        user: &ManagedAddress,
+        week: Week,
+    ) -> SingleValueMapper<bool>;
+}
+
+#[cfg(test)]
+mod claim_progress_tests {
```
Comment (psorinionut) [inline metabonding/src/claim_progress.rs:229]: Why have a test here?

Comment (dorin-iancu) [reply metabonding/src/claim_progress.rs:229]: It's a test for the struct's base logic, so there's not much reason to bundle it with the general contract tests.

**Review by psorinionut — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by ccorcoveanu — APPROVED** [review]
